### PR TITLE
[SHIP-3521] Add tests to L2EP contracts to improve test coverage

### PR DESCRIPTION
**Pull request by jlaveracll**
[SHIP-3521](https://smartcontract-it.atlassian.net/browse/SHIP-3521)

## This PR does:
- update minimum coverage % from 61 to 65
- since we unified some code into the shared `base` contracts, many tests were duplicated.
  - those tests have been removed, moved to the `base` contracts test cases, and for each implementation we now only test its specific overrides/implementations
- cleaned up some naming issues with tests and comments

## **Important**
I noticed that the `lcov_prune` script that removes the files that shouldn't be reported in the coverage report, was removing files that actually SHOULD be reportes, namely the `base` contracts implementations under the `/shared/` folder.

This happens because we intentionally and specifically want to exclude the files in the `/shared/` folder, but only those outside of the projects folders. The files in question here are in `/l2ep/dev/shared/`, which unfortunately also matches the `*/shared/*` expression.

In order to fix this, I had two options:
- update the `lcov_prune` file in order to prevent excluding the `l2ep/dev/shared` folder or to remove it from the exclusion list, or
- rename the `l2ep/dev/shared` folder to `l2ep/dev/base`

I opted to go for the later, as it would have a leaser impact on the overall project.
I'm open to different ideas and to change the approach if someone feels strongly about this or there's an issue with this implementation

### Requires Dependencies
Does not depend on any other PR

### Resolves Dependencies
No other PR depends on this one

[SHIP-3521]: https://smartcontract-it.atlassian.net/browse/SHIP-3521?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by chris-de-leon-cll — DISMISSED** [review]
This looks awesome!

It's great to finally see a simplified and refactored version of this code 💯 

Future L2EP code should be a lot easier to write once these changes get merged

File: contracts/src/v0.8/l2ep/test/v1_0_0/scroll/ScrollValidator.t.sol [hunk 1]
```diff
@@ -53,13 +53,19 @@ contract ScrollValidatorTest is L2EPTest {
...
-  /// @notice it correctly updates the gas limit
-  function test_CorrectlyUpdatesTheGasLimit() public {
-    uint32 newGasLimit = 2000000;
-    assertEq(s_scrollValidator.getGasLimit(), INIT_GAS_LIMIT);
-    s_scrollValidator.setGasLimit(newGasLimit);
-    assertEq(s_scrollValidator.getGasLimit(), newGasLimit);
+contract ScrollValidator_Constructor is ScrollValidatorTest {
+  /// @notice it should have been deployed with the correct initial state
+  function test_InitialState() public {
+    // Sets msg.sender and tx.origin to a valid address
```
Comment (RensR) [inline contracts/src/v0.8/l2ep/test/v1_0_0/scroll/ScrollValidator.t.sol]: imo no need for comments like these in tests. You can expect people to know what startPrank does. 

Also, I believe you only need to pass in the arg once, so you can remove the second argument

File: contracts/src/v0.8/l2ep/test/v1_0_0/scroll/ScrollValidator.t.sol [hunk 2]
```diff
@@ -53,13 +53,19 @@ contract ScrollValidatorTest is L2EPTest {
...
-contract ScrollValidator_SetGasLimit is ScrollValidatorTest {
-  /// @notice it correctly updates the gas limit
-  function test_CorrectlyUpdatesTheGasLimit() public {
-    uint32 newGasLimit = 2000000;
-    assertEq(s_scrollValidator.getGasLimit(), INIT_GAS_LIMIT);
-    s_scrollValidator.setGasLimit(newGasLimit);
-    assertEq(s_scrollValidator.getGasLimit(), newGasLimit);
+contract ScrollValidator_Constructor is ScrollValidatorTest {
+  /// @notice it should have been deployed with the correct initial state
+  function test_InitialState() public {
```
Comment (RensR) [inline contracts/src/v0.8/l2ep/test/v1_0_0/scroll/ScrollValidator.t.sol]: Please use the naming convention of 

`test_<function_being_tested>_<error_name(if any)_<success OR reverts>`

This is a bit harder with string errors as the name is very long. Maybe this ?

`test_constructor_InvalidL1MessageQueueAddress_reverts`?

File: contracts/src/v0.8/l2ep/test/v1_0_0/shared/BaseSequencerUptimeFeed.t.sol [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+    s_sequencerUptimeFeed.updateStatus(true, uint64(timestamp));
+    assertEq(s_sequencerUptimeFeed.latestAnswer(), 1);
+    assertEq(s_sequencerUptimeFeed.latestRound(), 2);
+    assertEq(s_sequencerUptimeFeed.latestTimestamp(), uint64(timestamp));
+
+    // Update with different status, but stale timestamp, should be ignored
+    timestamp = timestamp - 1000;
+    vm.expectEmit(false, false, false, false);
+    emit UpdateIgnored(true, 0, true, 0); // arguments are dummy values
+    // TODO: how can we check that an AnswerUpdated event was NOT emitted
```
Comment (RensR) [inline contracts/src/v0.8/l2ep/test/v1_0_0/shared/BaseSequencerUptimeFeed.t.sol]: You can use https://book.getfoundry.sh/cheatcodes/record-logs to record the logs and then assert none are emitted (or some other exact number you expect)

File: contracts/src/v0.8/l2ep/test/v1_0_0/shared/BaseSequencerUptimeFeed.t.sol [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+    // Sets msg.sender and tx.origin to a valid address
+    vm.startPrank(s_l1OwnerAddr, s_l1OwnerAddr);
+
+    // Gets data from a round that has not happened yet
+    vm.expectRevert(BaseSequencerUptimeFeed.NoDataPresent.selector);
+    s_sequencerUptimeFeed.getTimestamp(2);
+  }
+}
+
+contract BaseSequencerUptimeFeed_ProtectReadsOnAggregatorV2V3InterfaceFunctions is BaseSequencerUptimeFeedTest {
```
Comment (RensR) [inline contracts/src/v0.8/l2ep/test/v1_0_0/shared/BaseSequencerUptimeFeed.t.sol]: Contracts should be named `<contract_name>_<function_being_tested>`

File: contracts/src/v0.8/l2ep/test/v1_0_0/shared/BaseValidator.t.sol [hunk 5]
```diff
@@ -0,0 +1,56 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+
+import {BaseValidator} from "../../../dev/base/BaseValidator.sol";
+import {MockBaseValidator} from "../../mocks/MockBaseValidator.sol";
+import {L2EPTest} from "../L2EPTest.t.sol";
+
+contract BaseValidatorTest is L2EPTest {
+  address internal constant L2_SEQ_STATUS_RECORDER_ADDRESS = address(0x491B1dDA0A8fa069bbC1125133A975BF4e85a91b);
```
Comment (RensR) [inline contracts/src/v0.8/l2ep/test/v1_0_0/shared/BaseValidator.t.sol]: Are these addresses random? If so maybe nice to use `makeAddr("L2_SEQ_STATUS_RECORDER_ADDRESS");` to clearly indicate we don't care about the actual value

File: contracts/src/v0.8/l2ep/test/v1_0_0/scroll/ScrollValidator.t.sol (hunk 1 above)
Comment (jlaveracll) [reply contracts/src/v0.8/l2ep/test/v1_0_0/scroll/ScrollValidator.t.sol]: Cleaned up the comments

> If the alternative signature of startPrank is used, then tx.origin is set as well for all subsequent calls.

for just two test cases this is necessary, for the rest I removed it

File: contracts/src/v0.8/l2ep/test/v1_0_0/shared/BaseValidator.t.sol (hunk 5 above)
Comment (jlaveracll) [reply contracts/src/v0.8/l2ep/test/v1_0_0/shared/BaseValidator.t.sol]: looks much better! updated

File: contracts/src/v0.8/l2ep/test/v1_0_0/shared/BaseSequencerUptimeFeed.t.sol (hunk 3 above)
Comment (jlaveracll) [reply contracts/src/v0.8/l2ep/test/v1_0_0/shared/BaseSequencerUptimeFeed.t.sol]: great!

File: contracts/src/v0.8/l2ep/test/v1_0_0/shared/BaseSequencerUptimeFeed.t.sol (hunk 4 above)
Comment (mohamed-mehany) [reply contracts/src/v0.8/l2ep/test/v1_0_0/shared/BaseSequencerUptimeFeed.t.sol]: Fixed!

File: contracts/src/v0.8/l2ep/test/v1_0_0/scroll/ScrollValidator.t.sol (hunk 2 above)
Comment (mohamed-mehany) [reply contracts/src/v0.8/l2ep/test/v1_0_0/scroll/ScrollValidator.t.sol]: Fixed!

**Review by chris-de-leon-cll — DISMISSED** [review]

**Review by RensR — APPROVED** [review]

**Review by chris-de-leon-cll — APPROVED** [review]

**Review by erikburt — APPROVED** [review]

**Review by prashantkumar1982 — APPROVED** [review]

**Pull request merged**